src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +167,268 @@ func (b *dbBuffer) Write(
 	value float64,
 	unit xtime.Unit,
 	annotation []byte,
+	wopts WriteOptions,
 ) (bool, error) {
 	now := b.nowFn()
-	futureLimit := now.Add(1 * b.bufferFuture)
-	pastLimit := now.Add(-1 * b.bufferPast)
-	if !futureLimit.After(timestamp) {
-		return false, m3dberrors.ErrTooFuture
-	}
-	if !pastLimit.Before(timestamp) {
-		return false, m3dberrors.ErrTooPast
-	}
+	wType := wopts.ResolveWriteType(timestamp, now, b.bufferPast, b.bufferFuture)
 
-	bucketStart := timestamp.Truncate(b.blockSize)
-	idx := b.writableBucketIdx(timestamp)
-	if b.buckets[idx].needsReset(bucketStart) {
-		// Needs reset
-		b.DrainAndReset()
+	if wType == ColdWrite {
+		if !b.coldWritesEnabled {
+			return false, m3dberrors.ErrColdWritesNotEnabled
+		}
+
+		if now.Add(-b.retentionPeriod).Truncate(b.blockSize).After(timestamp) {
```

Comment:
Hm, I think even if cold writes are enabled you want to only allow `now-retention` not `(now-retention).truncate(blocksize)`.

You still should only be able to write into the past as far back as retention is enabled for your namespace.

Reply:
Makes sense.a

Reply:
Still need to take these truncate statements out, just FYI.